cmd/lotus-shed/f3.go

```diff
@@ -61,6 +72,103 @@ func loadF3IDList(path string) ([]gpbft.ActorID, error) {
 	return ids, nil
 }
 
+var f3CheckActivation = &cli.Command{
+	Name: "check-activation",
+
+	Flags: []cli.Flag{
+		&cli.StringFlag{
+			Name: "contract",
+		},
+	},
+	Action: func(cctx *cli.Context) error {
+		address, err := ethtypes.ParseEthAddress(cctx.String("contract"))
+		if err != nil {
+			return fmt.Errorf("trying to parse contract address: %s: %w", cctx.String("contract"), err)
+		}
+
+		ethCall := ethtypes.EthCall{
+			To:   &address,
+			Data: must.One(ethtypes.DecodeHexString("0x2587660d")), // method ID of activationInformation()
+		}
+		fMessage, err := ethCall.ToFilecoinMessage()
+		if err != nil {
+			return fmt.Errorf("converting to filecoin message: %w", err)
+		}
+
+		ctx := cliutil.ReqContext(cctx)
+		api, closer, err := cliutil.GetFullNodeAPIV1(cctx)
+		if err != nil {
+			return fmt.Errorf("getting api: %w", err)
+		}
+		defer closer()
+
+		msgRes, err := api.StateCall(ctx, fMessage, types.EmptyTSK)
+		if err != nil {
+			return fmt.Errorf("state call error: %w", err)
+		}
+		if msgRes.MsgRct.ExitCode != 0 {
+			return fmt.Errorf("message returned exit code: %v", msgRes.MsgRct.ExitCode)
+		}
+
+		var ethReturn abi.CborBytes
+		err = ethReturn.UnmarshalCBOR(bytes.NewReader(msgRes.MsgRct.Return))
+		if err != nil {
+			return fmt.Errorf("could not decode return value: %w", err)
+		}
+		slot, retBytes := []byte{}, []byte(ethReturn)
+		// 3*32 because there should be 3 slots minimum
+		if len(retBytes) < 3*32 {
```

Comment:
Would be good to refactor magic contract checks like this into one file with meaningfully named functions for better code readability.

Reply:
This is getting into realm of more generic decoders which exist but it was easier not to use them.